src/solvers/strings/string_constraint_generator_constants.cpp

```diff
@@ -78,6 +78,7 @@ add_axioms_for_empty_string(const function_application_exprt &f)
 /// \param arg: expression of type string typet
 /// \param guard: condition under which `res` should be equal to arg
 /// \param array_pool: pool of arrays representing strings
+/// \param ns: the namespace
```

Comment:
:bulb: Maybe it will be good to say that this argument is so that we can simplify some expressions. We could pass a function `simplify_expr` as argument instead to make that obvious (just an idea)

Reply:
Or maybe we could have a class and have some const members (such as a `namespacet`)? The architecture of the string solver code makes it really painful to add new capabilities.

Reply:
We could basically use the `string_constraint_generatort` class for that (either pass it as an argument of the function or make the function a member), but we should clean-up the members of the class first: the `string_constraintst` field should just be an output of the methods and not a field and I think the `messaget` field does not belong there.

Reply:
I don't know what exactly the right solution looks like, but I get the impression that making everything free functions has been taken a little too far here. It severely hampers maintainability. The PR we are looking at could have been a one-line change, but instead it's currently +87/-48. We need to make sure we find a good balance between perfectly clean code and maintainability.

Reply:
Here is the clean-up I suggest https://github.com/diffblue/cbmc/pull/4995, you could then replace the `fresh_symbol`, `array_pool`, and `ns` arguments in these functions by just a string_constraint_generatort object and then further changes like this one could be small.

Reply:
I'll happily do that. I was tempted to do it originally while doing this, but the changes were just under the acceptable threshold for me.